File: mobile/tests/features/intros/warmIntros.service.test.ts
jest.mock('~/lib/supabase/client', () => ({
  supabase: { rpc: jest.fn() },
}));

import { supabase } from '~/lib/supabase/client';
import {
  suggestWarmIntros,
  sendWarmRequest,
  forwardWarmIntro,
} from '~/features/intros/services/warmIntros.service';
import {
  IntroDuplicateError,
  IntroError,
  IntroRateLimitError,
} from '~/features/intros/services/intros.service';

describe('warmIntros.service', () => {
  beforeEach(() => jest.clearAllMocks());

  describe('suggestWarmIntros', () => {
    it('calls suggest_warm_intros RPC with the default limit and maps rows to camelCase DTOs', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({
        data: [
          {
            target_id: 't1',
            target_handle: 'bob',
            target_name: 'Bob',
            target_photo_url: null,
            target_primary_role: 'founder',
            target_goal_type: 'co_found',
            mutual_count: 2,
            top_mutual_id: 'm1',
            top_mutual_name: 'Alice',
            top_mutual_handle: 'alice',
          },
        ],
        error: null,
      });
      const result = await suggestWarmIntros();
      expect(supabase.rpc).toHaveBeenCalledWith('suggest_warm_intros', { p_limit: 10 });
      expect(result).toEqual([
        {
          targetId: 't1',
          targetHandle: 'bob',
          targetName: 'Bob',
          targetPhotoUrl: null,
          targetPrimaryRole: 'founder',
          targetGoalType: 'co_found',
          mutualCount: 2,
          topMutualId: 'm1',
          topMutualName: 'Alice',
          topMutualHandle: 'alice',
        },
      ]);
    });

    it('honours an explicit limit', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({ data: [], error: null });
      await suggestWarmIntros(25);
      expect(supabase.rpc).toHaveBeenCalledWith('suggest_warm_intros', { p_limit: 25 });
    });

    it('returns [] when the RPC returns null data', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({ data: null, error: null });
      expect(await suggestWarmIntros()).toEqual([]);
    });

    it('throws when the RPC returns an error', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({
        data: null,
        error: { message: 'unauthenticated' },
      });
      await expect(suggestWarmIntros()).rejects.toThrow('unauthenticated');
    });
  });

  describe('sendWarmRequest', () => {
    it('calls send_warm_request RPC with snake_case params and returns the new id', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({ data: 'new-intro-id', error: null });
      const id = await sendWarmRequest({
        mutualId: 'm1',
        targetId: 't1',
        note: 'x'.repeat(100),
      });
      expect(supabase.rpc).toHaveBeenCalledWith('send_warm_request', {
        p_mutual_id: 'm1',
        p_target_id: 't1',
        p_note: 'x'.repeat(100),
      });
      expect(id).toBe('new-intro-id');
    });

    it('maps 23505 errors onto IntroDuplicateError', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({
        data: null,
        error: { code: '23505', message: 'warm request already pending' },
      });
      await expect(
        sendWarmRequest({ mutualId: 'm1', targetId: 't1', note: 'x'.repeat(100) })
      ).rejects.toBeInstanceOf(IntroDuplicateError);
    });

    it('maps daily-cap P0001 onto IntroRateLimitError', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({
        data: null,
        error: { code: 'P0001', message: 'daily cap reached' },
      });
      await expect(
        sendWarmRequest({ mutualId: 'm1', targetId: 't1', note: 'x'.repeat(100) })
      ).rejects.toBeInstanceOf(IntroRateLimitError);
    });

    it('maps other Postgrest errors onto the generic IntroError', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({
        data: null,
        error: { code: '42501', message: 'no connection to mutual' },
      });
      await expect(
        sendWarmRequest({ mutualId: 'm1', targetId: 't1', note: 'x'.repeat(100) })
      ).rejects.toBeInstanceOf(IntroError);
    });
  });

  describe('forwardWarmIntro', () => {
    it('calls forward_warm_intro RPC with snake_case params and returns the new id', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({ data: 'forward-id', error: null });
      const id = await forwardWarmIntro({
        introId: 'i1',
        note: 'y'.repeat(120),
      });
      expect(supabase.rpc).toHaveBeenCalledWith('forward_warm_intro', {
        p_intro_id: 'i1',
        p_note: 'y'.repeat(120),
      });
      expect(id).toBe('forward-id');
    });

    it('throws IntroError on RPC failure', async () => {
      (supabase.rpc as jest.Mock).mockResolvedValueOnce({
        data: null,
        error: { code: '22023', message: 'warm request not in delivered state' },
      });
      await expect(
        forwardWarmIntro({ introId: 'i1', note: 'y'.repeat(120) })
      ).rejects.toBeInstanceOf(IntroError);
    });
  });
});

File: mobile_flutter/test/features/intros/data/warm_intros_service_test.dart
import 'package:connect_mobile/core/errors/app_exception.dart';
import 'package:connect_mobile/features/intros/data/warm_intros_service.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../../helpers/intros_fixtures.dart';

class _FakeWarmGateway extends Mock implements WarmIntrosGateway {}

void main() {
  setUpAll(() => registerFallbackValue(<String, dynamic>{}));

  group('suggestWarmIntros', () {
    test('returns list of WarmSuggestion with passed limit', () async {
      final gateway = _FakeWarmGateway();
      final service = WarmIntrosService(gateway);
      final s = buildWarmSuggestion();
      when(
        () => gateway.rpc('suggest_warm_intros', params: any(named: 'params')),
      ).thenAnswer((_) async => <Map<String, dynamic>>[s.toJson()]);

      final result = await service.suggestWarmIntros(limit: 5);
      expect(result, hasLength(1));
      expect(result.single.targetId, s.targetId);
      final captured = verify(
        () => gateway.rpc(
          'suggest_warm_intros',
          params: captureAny(named: 'params'),
        ),
      ).captured.single as Map<String, dynamic>;
      expect(captured['p_limit'], 5);
    });

    test('default limit is 10', () async {
      final gateway = _FakeWarmGateway();
      final service = WarmIntrosService(gateway);
      when(
        () => gateway.rpc('suggest_warm_intros', params: any(named: 'params')),
      ).thenAnswer((_) async => <Map<String, dynamic>>[]);

      await service.suggestWarmIntros();
      final captured = verify(
        () => gateway.rpc(
          'suggest_warm_intros',
          params: captureAny(named: 'params'),
        ),
      ).captured.single as Map<String, dynamic>;
      expect(captured['p_limit'], 10);
    });

    test('maps 28000 -> UnauthenticatedException', () async {
      final gateway = _FakeWarmGateway();
      final service = WarmIntrosService(gateway);
      when(
        () => gateway.rpc('suggest_warm_intros', params: any(named: 'params')),
      ).thenThrow(const PostgrestException(message: '', code: '28000'));
      expect(
        () => service.suggestWarmIntros(),
        throwsA(isA<UnauthenticatedException>()),
      );
    });
  });

  group('sendWarmRequest', () {
    test('trims note and returns new intro id', () async {
      final gateway = _FakeWarmGateway();
      final service = WarmIntrosService(gateway);
      when(
        () => gateway.rpc('send_warm_request', params: any(named: 'params')),
      ).thenAnswer((_) async => 'new-intro-id');

      final id = await service.sendWarmRequest(
        mutualId: 'm',
        targetId: 't',
        note: '  ${'a' * 100}  ',
      );
      expect(id, 'new-intro-id');
      final captured = verify(
        () => gateway.rpc(
          'send_warm_request',
          params: captureAny(named: 'params'),
        ),
      ).captured.single as Map<String, dynamic>;
      expect(captured['p_mutual_id'], 'm');
      expect(captured['p_target_id'], 't');
      expect((captured['p_note'] as String).length, 100);
    });

    test('maps 23505 -> DuplicateException (anti-shotgun)', () async {
      final gateway = _FakeWarmGateway();
      final service = WarmIntrosService(gateway);
      when(
        () => gateway.rpc('send_warm_request', params: any(named: 'params')),
      ).thenThrow(const PostgrestException(message: '', code: '23505'));
      expect(
        () => service.sendWarmRequest(
          mutualId: 'm',
          targetId: 't',
          note: 'x' * 100,
        ),
        throwsA(isA<DuplicateException>()),
      );
    });
  });

  group('forwardWarmIntro', () {
    test('trims note and returns new forward id', () async {
      final gateway = _FakeWarmGateway();
      final service = WarmIntrosService(gateway);
      when(
        () => gateway.rpc('forward_warm_intro', params: any(named: 'params')),
      ).thenAnswer((_) async => 'forward-id');

      final id = await service.forwardWarmIntro(
        introId: 'wr-1',
        note: '  ${'b' * 90}  ',
      );
      expect(id, 'forward-id');
      final captured = verify(
        () => gateway.rpc(
          'forward_warm_intro',
          params: captureAny(named: 'params'),
        ),
      ).captured.single as Map<String, dynamic>;
      expect(captured['p_intro_id'], 'wr-1');
      expect((captured['p_note'] as String).length, 90);
    });

    test('maps 22023 note-range -> IntroNoteRangeException', () async {
      final gateway = _FakeWarmGateway();
      final service = WarmIntrosService(gateway);
      when(
        () => gateway.rpc('forward_warm_intro', params: any(named: 'params')),
      ).thenThrow(
        const PostgrestException(
          message: 'char_length(btrim(note))',
          code: '22023',
        ),
      );
      expect(
        () => service.forwardWarmIntro(introId: 'i', note: 'short'),
        throwsA(isA<IntroNoteRangeException>()),
      );
    });
  });
}
